DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1

```diff
@@ -521,6 +518,119 @@ function Get-ChocoInstalledPackage {
     Return $res
 }
 
+<#
+.Synopsis
+   Run Chocolatey executable and throws error on failure
+.DESCRIPTION
+   Run Chocolatey executable and throws error on failure
+.EXAMPLE
+   Invoke-Chocolatey "list -lo"
+.EXAMPLE
+   Invoke-Chocolatey -arguments "list -lo"
+#>
+function Invoke-Chocolatey {
+    [CmdletBinding()]
+    Param
+    (
+        # Chocolatey arguments."
+        [Parameter(Position = 0)]
+        [string]$arguments
+    )
+
+    $result = Invoke-ChocoProcess -arguments $arguments
+
+    #exit codes reference: https://docs.chocolatey.org/en-us/choco/commands/install#exit-codes
+    switch ($result.exitcode) {
+        0 {
+            #most widely used success exit code
+            $result.output.Split("`n")
+            break
+        }
+        350 {
+            # pending reboot detected, no action has occurred
+            $result.output.Split("`n")
+            throw "Error: Chocolatey detected a pending reboot from a previous installation. You can modify your DSC and use the PendingReboot DSC resource to reboot before running this command."
+            break
+        }
+        1605 {
+            #(MSI uninstall) - the product is not found, could have already been uninstalled
+            $result.output.Split("`n")
+            break
+        }
+        1614 {
+            #(MSI uninstall) - the product is uninstalled
+            $result.output.Split("`n")
+            break
+        }
+        1641 {
+            #(MSI) - restart initiated
+            $result.output.Split("`n")
+            Request-DscReboot
+            break
+        }
+        3010 {
+            #(MSI, InnoSetup can be passed to provide this) - restart required
+            $result.output.Split("`n")
+            Request-DscReboot
+            break
+        }
+        default {
+            #when error, throw output as error, contains errormessage
+            throw "Error: Chocolatey command failed with exit code $($result.exitcode).`n$($result.output)" 
+        }
+    }
+}
+
+function Invoke-ChocoProcess {
+    [CmdletBinding()]
+    [OutputType([hashtable])]
+    Param
+    (
+        # Chocolatey arguments."
+        [Parameter(Position = 0)]
+        [string]$arguments
+    )
+
+    Write-Verbose -Message "command: 'choco $arguments'"
+
+    $pinfo = New-Object System.Diagnostics.ProcessStartInfo
+    $pinfo.FileName = 'choco'
+    $pinfo.RedirectStandardError = $true
+    $pinfo.RedirectStandardOutput = $true
+    $pinfo.UseShellExecute = $false
+    $pinfo.Arguments = "$arguments"
+
+    $p = New-Object System.Diagnostics.Process
+    $p.StartInfo = $pinfo
+    $p.Start() | Out-Null
+
+    $output = $p.StandardOutput.ReadToEnd()
+    $p.WaitForExit()
+    $exitcode = $p.ExitCode
+    $p.Dispose()
+
+    #Set $LASTEXITCODE variable.
+    powershell.exe -NoLogo -NoProfile -Noninteractive "exit $exitcode"

```

Comment:
This PR is basically a "port" of the old PR #103 from 2017 to the latest code base so I cannot tell you exactly why they did it that way.

The only reason I can see to do it that way is to ensure that the real `$LastExitCode` is set. Setting `$LastExitCode` directly would create a variable named `$LastExitCode` which is scoped to the function `Invoke-ChocoProcess` and it would not set the global `$LastExitCode` variable.

You can reproduce the behavior with this code

```powershell
function Test {
    $LASTEXITCODE = 3 # this will create a variable scoped to the function and will not impact its parent scope

   # you can uncomment this line to set the global $LastExitCode 
   # powershell.exe -NoLogo -NoProfile -Noninteractive "exit 3"
}

#fake a lastexitcode
powershell.exe -NoLogo -NoProfile -Noninteractive "exit 123"

write-Host "Value before Test function: $LASTEXITCODE" # will print 123 for the exit code (set by the previous line)
Test
write-Host "Value after Test function: $LASTEXITCODE" # value stays at 123 because the exit code variable in the Test function is scope to that function
```